### PR TITLE
Fix Hash#from_xml with frozen strings

### PR DESCRIPTION
**Pull request by marekkirejczyk**
Fix Hash#from_xml with frozen strings as an argument, so it won't raise IOError: not modifiable string.

### CODE REVIEW

File: activesupport/lib/active_support/xml_mini/rexml.rb [hunk 1]
```diff
@@ -20,11 +20,9 @@ def parse(data)
         data = StringIO.new(data || '')
       end
 
-      char = data.getc
-      if char.nil?
+      if data.eof?
```
Comment (vipulnsward) [inline activesupport/lib/active_support/xml_mini/rexml.rb]: Does `data.eof?` mean we have no data?
Some data might already have been read and we might falsely consider it as eof? ?

Comment (marekkirejczyk) [inline activesupport/lib/active_support/xml_mini/rexml.rb]: It means there is no more data to read, detalis here: [IO#eof?](http://ruby-doc.org/core-2.3.0/IO.html#method-i-eof-3F). 
Returns true if getc returns nil [getc](http://ruby-doc.org/core-2.3.0/IO.html#method-i-getc), but will not eat a char.

**Comment by marekkirejczyk** [comment]
I did more digging and noticed that rexml tests are week and try to compare function result with itself (see [assert_equal_rexml](https://github.com/rails/rails/blob/master/activesupport/test/xml_mini/rexml_engine_test.rb#L29). 

To make sure it is the case I changed parse function to always return the same value and the test would still pass (see [this gist](https://gist.github.com/marekkirejczyk/6c89d7efbb0b479a729b4bea04ee9c05)).

Commit with fixed test attached. 

I also noticed that frozen string issue most likely present in libxml, nokogiri and both sax parsers as code looks to be copy passed. Working on patch.

**Comment by rafaelfranca** [comment]
Backported in 90c5e3d and d8d23f3

**Pull request merged**